### PR TITLE
setup announcement actor, add endpoint for actor

### PR DESCRIPTION
**Pull request by catdevnull**
Leaving this here so it can be reviewed as I go

ref 
* https://github.com/hyphacoop/distributed-press-organizing/issues/101
* https://github.com/hyphacoop/distributed-press-organizing/issues/122
* https://github.com/hyphacoop/distributed-press-organizing/issues/123
* https://github.com/hyphacoop/distributed-press-organizing/issues/124

### CODE REVIEW

**Review by RangerMauve — COMMENTED** [review]
This is a good start. The AP stuff looks correct. we just need the inbox now so we can accept and track followers.

File: src/server/api/announcements.ts
Comment (RangerMauve) [inline src/server/api/announcements.ts:45]: Yeah that's fine

File: src/server/api/creation.ts
Comment (RangerMauve) [inline src/server/api/creation.ts:32]: Instead could you add a `actorsDB.has(actor)` API? This seems kinda inefficient.

File: src/server/api/creation.ts
Comment (RangerMauve) [inline src/server/api/creation.ts:42]: Could you add a new class to contain all this stuff and attach it to `APSystem.announcements`? There should be a method like `apsystem.announcements.announce(actor)` which can perform activity creation.

File: src/server/api/creation.ts
Comment (RangerMauve) [inline src/server/api/creation.ts:56]: Would the actor need to be `announcements` in this case? Announcing as the newly created actor doesn't make sense since they have no followers.

File: src/server/api/index.ts
Comment (RangerMauve) [inline src/server/api/index.ts:108]: What's the purpose for re-invoking `setInfo` is it in case the publicURL changes? Might be good to only invoke it inside `catch` or if `!prev.actorURL === actorURL`. Also per the previous comment, mind putting this logic into `apsystem.announcements.init(publicURL)`

File: src/server/store/index.ts
Comment (RangerMauve) [inline src/server/store/index.ts:40]: Maybe we should change this to `@announcements@{new URL(publicURL).host}`? Then we can avoid some of the trouble with expecting the `:actor` to be in `@username@hostname` format.

File: src/scripts/add-post-to-announcements.ts
Comment (RangerMauve) [inline src/scripts/add-post-to-announcements.ts:1]: We should probably remove this script before merging.

**Review by RangerMauve — CHANGES_REQUESTED** [review]
We also need to implement the `webfinger` protocol. Specifically we need to resolve `/.well-known/webfinger` to a json file with the mime type [`application/jrd+json`](https://github.com/hyphacoop/api.distributed.press/blob/main/ansible/roles/distributed_press/templates/nginx-static.j2#L22).

example: https://github.com/RangerMauve/staticpub.mauve.moe/blob/default/.well-known/webfinger

This is necessary to enable AP instances to resolve the actor. We should use `announcements@{hostname of the server}` as the account others would follow. e.g. `@announcements@social.distributed.press`

File: src/client/cli.ts
Comment (catdevnull) [reply src/client/cli.ts:10]: the reason I didn't use the .command API yet is because I don't think it's possible to set up as `cli <email> <command>` and i figured that because this is an internal tool, leaving it without that implemented yet until it's needed that it does more stuff is ok. check out the issue i linked.

but.. i'm not strongly opinionated about the CLI. if you can suggest a CLI that better adapts to what yargs provides i'll happily implement it

File: src/scripts/add-post-to-announcements.ts
Comment (catdevnull) [reply src/scripts/add-post-to-announcements.ts:1]: couldn't it be useful (maybe implemented in a proper way) to post stuff to the announcements actor manually?

even then, it can be useful for debugging in staging. is staging strictly the `main` branch or can it include these sort of changes?

File: src/server/api/creation.ts
Comment (catdevnull) [reply src/server/api/creation.ts:32]: i tried that, but at least my editor tells me that `.has` doesn't exist in abstract-level. `.get` returns `any` ( :( ), so i'm not really sure how to do this "properly"

File: src/server/store/index.ts
Comment (catdevnull) [reply src/server/store/index.ts:40]: hmm, i'm not too sure about this. that would mean that there wouldn't be a "special" announcements ActorStore in the global store, right?

File: src/client/cli.ts
Comment (RangerMauve) [reply src/client/cli.ts:10]: I think it'd be fine if the email was passed in as an environment variable or a command line flag, personally. I found the `.command` syntax to be convenient to use for other projects.

File: src/server/api/creation.ts
Comment (RangerMauve) [reply src/server/api/creation.ts]: Ah, you can perform a `.get` and check that the value is not undefined. https://www.npmjs.com/package/abstract-level#dbgetkey-options

Or I think actorsDb.get throws an error on not found so maybe .has can catch that and return false

File: src/server/store/index.ts
Comment (RangerMauve) [reply src/server/store/index.ts:40]: Oh, I didn't mean to remove the special actor store, just to have the name for it follow the usual `@user@host` syntax.

File: src/scripts/add-post-to-announcements.ts
Comment (RangerMauve) [reply src/scripts/add-post-to-announcements.ts:1]: I'm not sure actually. Having it be an admin-only HTTP API (POST `/v1/@announcements@social.distributed.press/outbox`) could be useful for that. It's just that scripts like this can only run when the server is shut down so we should be very sparing with how much functionality requires that. So far I think they've just been there for initial server setup.

File: src/server/api/announcements.ts
Comment (RangerMauve) [inline src/server/api/announcements.ts:70]: Could you move the post listing logic into the announcements class as well?

**Review by RangerMauve — CHANGES_REQUESTED** [review]
Excellent progress, just some minor bits left.

File: src/server/store/index.ts
Comment (catdevnull) [reply src/server/store/index.ts:40]: okay, this makes sense, but that also implies:
1. adding announcements to the global actors `store`
   
   sidenote: i noticed that `this.actorsDb` isn't being used for anything, and actual actors are just being written as a sublevel to the "global" DB. is that intended behavior?
2. having "special" endpoints for this snowflake actor. i don't know how fastify handles that; do we need to define them before the generic `/:actor/` endpoints? also, how do we define the routes with the publicURL? do we just define them "dynamically" using cfg.publicURL?

imo having the snowflake case be just here makes more sense, but i'm not strongly opiniated

File: src/scripts/add-post-to-announcements.ts
Comment (catdevnull) [reply src/scripts/add-post-to-announcements.ts:1]: well, i didn't implement the POST /outbox as we decided that was out of scope lol

so uhh, your call, do i make the CLI changes to keep the script around or not?

File: src/server/store/index.ts
Comment (RangerMauve) [reply src/server/store/index.ts:40]: > adding announcements to the global actors store

We don't need to do this since we have this if statement. We're using `forActor` to get any actors so there's no other path the announcements store would be fetched as an actor by accident. The actosDB map is just a cache for the sublevel instances so that we don't needlessly initialize them.

> imo having the snowflake case be just here makes more sense, but i'm not strongly opiniated

Mind elaborating on this point?

> do we need to define them before the generic /:actor/ endpoints? also, how do we define the routes with the publicURL? do we just define them "dynamically" using cfg.publicURL?

Yes to all

File: src/scripts/add-post-to-announcements.ts
Comment (RangerMauve) [reply src/scripts/add-post-to-announcements.ts:1]: Lets use it for testing for now then remove before release and reivist again with the post to outbox use case.

File: src/server/store/index.ts
Comment (RangerMauve) [reply src/server/store/index.ts:40]: Hmm, thinking about this more, how about we don't try to shove announcements into `forActor` and instead keep the `/v1/announcements/` section as is and add a dedicated inbox. It should just auto accept follows and undos on follows if there's no global blocklist for them and ignore all else for now.

File: src/server/announcements.ts
Comment (RangerMauve) [inline src/server/announcements.ts:30]:
```suggestion
  getActor (): Promise<ActorStore> {
```

frActor is sync so this should be too

File: src/server/announcements.ts
Comment (RangerMauve) [inline src/server/announcements.ts:76]:
```suggestion
      await this.getActor().outbox.add(activity)
```

**Review by RangerMauve — CHANGES_REQUESTED** [review]

**Comment by catdevnull** [comment]
* [ ] add unit test for checking that after creating a new inbox/user with announce: true, an announcement appears in the announcement account

**Comment by RangerMauve** [comment]
Getting an error related to sublevel not found when testing

File: src/server/announcements.ts
Comment (RangerMauve) [inline src/server/announcements.ts:64]: Why would undefined mean that it exists already? Wouldn't it mean that it doesn't currently exist? Maybe we should catch both that and the level error and throw an error for both.

**Review by RangerMauve — CHANGES_REQUESTED** [review]

**Comment by RangerMauve** [comment]
Deploying to staging!

File: src/server/api/creation.ts
Comment (RangerMauve) [inline src/server/api/creation.ts:38]: Since we're setting the actor before that, announce will never see it as new.

**Review by RangerMauve — CHANGES_REQUESTED** [review]

File: src/server/announcements.ts
Comment (RangerMauve) [inline src/server/announcements.ts:63]: This would only check if the actor's DB is in memory. To check if an actor exists we should use `.forActor(actor).getInfo()`

**Review by RangerMauve — CHANGES_REQUESTED** [review]
Maybe a new API should be made to check if the actor is existing, then set info, then announce if it was new. Place the logic outside of announcements and outside of the http API. Like `server/creation.ts`?

File: src/server/announcements.ts
Comment (RangerMauve) [inline src/server/announcements.ts:76]: This needs to be the announcements actorUrl
```suggestion
        actor: this.actorUrl,
```

File: src/server/announcements.ts
Comment (RangerMauve) [inline src/server/announcements.ts:74]: ActivityStreams hold activities, not Notes. This should be wrapped in a Create activity with the Note being the `object`.

e.g. here's how we link to notes in the staticpub example: https://github.com/RangerMauve/staticpub.mauve.moe/blob/default/outbox.jsonld#L13

We may need to store the notes in a separate store and add a new `/:actor/note/:id` URL to resolve them.

**Review by RangerMauve — CHANGES_REQUESTED** [review]

File: src/server/announcements.ts
Comment (catdevnull) [reply src/server/announcements.ts:74]: I don't see a requirement in the [spec](https://www.w3.org/TR/activitystreams-core/) that states that OrderedCollections must hold Activitys. Actually, I was mostly copying off of Sutty's Jekyll implementation, which has plain Updates inside the outbox:
* https://sutty.nl/outbox.jsonld -> https://sutty.nl/lanzamiento-de-publicaciones-distribuidas-en-el-fediverso-a-trav%C3%A9s-de-sutty/update/1701726785.jsonld

Am I getting something wrong?

Comment (RangerMauve) [reply src/server/announcements.ts:74]: Yeah, notice how the update you linked wraps the note in an Update activity. We should do the same here, but with a Create.

File: src/server/announcements.ts
Comment (RangerMauve) [inline src/server/announcements.ts:77]: The attributedTo should also be the announcements actor.

File: src/server/api/creation.ts
Comment (RangerMauve) [inline src/server/api/creation.ts:38]: I worry this brings a race condition where we announce the actor exists before it's fully registered. We should detect if actor is new outside of the `announce` method and set the info before making the announcement.

**Review by RangerMauve — CHANGES_REQUESTED** [review]

File: src/server/announcements.ts
Comment (RangerMauve) [reply src/server/announcements.ts:74]: We might, lets see if just wrapping is enough for now though.

File: src/server/announcements.ts
Comment (RangerMauve) [inline src/server/announcements.ts:74]: I think we can change this to be `#note` then the json-ld magic will resolve it? I'll test like this first.

File: src/server/announcements.ts
Comment (catdevnull) [inline src/server/announcements.ts:125]: are we supposed to have both the Create and the Note in the outbox?

File: src/server/announcements.ts
Comment (catdevnull) [inline src/server/announcements.ts]: this should probably be moved elsewhere but i'm not sure where. maybe src/schemas.ts? either way it can be moved later if it's used elsewhere

File: src/server/announcements.ts
Comment (catdevnull) [reply src/server/announcements.ts]: I know nothing about json-ld :shrug: but sounds good to me

**Review by catdevnull — COMMENTED** [review]
looks good!

File: src/server/announcements.ts
Comment (RangerMauve) [reply src/server/announcements.ts]: Yeah mind moving it to schemas? Also maybe we could PR the field into the module? They accepted my orderedItems one

Comment (catdevnull) [reply src/server/announcements.ts]: i'm looking into doing the PR but I can't find the spec that defines keypair in APActor

**Pull request merged**
